### ISSUE TITLE
Are contributions welcome?

### ISSUE THREAD

**Issue by cristi-badila** [steve-hyperbolic/hyper#4]
Hi Steve, my name is Cristian and as "hinted" to by the title, my first question is: are you open to merging code from other people?

If yes, I have a second question: have you considered moving each of the projects into separate repositories? 

I ask the second question as I found HyperMock a couple of months back and thought it looked promising as I needed a mocking library for a UWP project I'm working on and off again. Unfortunately it did not support all of the features I wanted and so I thought I could contribute to it. When I landed here on GitHub, to my surprise I found some other projects bundled in the same repository. This makes contributing to the project a bit more cumbersome and would also make organizing issues/tickets for a specific project much more difficult. As such I've been working on my own fork trying to make the library more extensible and increasing the test coverage (feel free to take a look [here](https://github.com/cristi-badila/hyper)). However I wouldn't necessarily want to actually fork the project and publish it under another name. Instead I was thinking if we could talk about the implementation/collaborate on it and maybe move along in the same direction.

**Comment by steve-hyperbolic** [comment]
Hi Cristian

Thanks for taking the time to use and expand the HyperMock framework.

Before I answer your questions, let me give you a brief history of the code in the repository. 

Basically I never really expected any of it to get used by others. I use it for a variety of personal projects. Some of the code has been around for years such as HyperIoC - albeit evolved over time. HyperMock actually came from some old code I wrote about 4 years ago where I built a whole unit testing framework with mocking built in, along with its own proxy generation. It worked well but I never got the framework fully integrated into test runners such as R#. It did have its own built-in version mind.

For the HyperMock project, I have already had one or two contributions from others and I am also re-shuffling the internals a bit too. I appreciate the code for HyperMock is a little rubbish in places. 

I want to be careful with respect to making breaking changes. Does your contribution change the original public API? I will take a look at your fork.

I have no issue with creating a new repo for HyperMock. Given the interest from others then it would make sense to do that. 

The bottom line is yes I am happy for contributions but just need to be careful how we go about it. I will review your fork and stick another message on here shortly.

Kind regards
Steve

**Comment by cristi-badila** [comment]
Thanks for the quick reply Steve. At the moment the state of the fork would indeed introduce some breaking changes, the most important of which would be that when creating a mock for type T, instead of obtaining directly an instance of ProxyDispatcher, you get a Mock<T> instance. This was so as to not pollute the global namespace with the Setup and Verify extension methods. In regards to the public API there are one or more minor breaking changes but I think it would be best if we could have a call/hangout at some point and discuss if you consider the changes I've made have merit, and if/how would be best to introduce them. My Skype id is: `cristian-badila`. Let me know when you would have the time to talk.

**Comment by steve-hyperbolic** [comment]
Hi Cristian

I have taken a look at your fork. The code design looks a lot cleaner than my current version. There are a few things we need to consider. In no particular order:

- Mock.Object - this would be a breaking change to get the instance. Given the implementation I am not sure there are any real options to address this (or even want to)
- Param class changed to Is. Again this is a breaking change. We may need to re-add it with an obsolete attribute or just bite the bullet and break peoples code on the next release. Not keen to do the latter.
- Public classes. In my original design, I made use of internal classes and the InternalsVisibleTo to provide access to certain classes that I would otherwise not want made public. An example is the Dispatcher property on the Mock class. I am not convinced this nor the associated classes actually want to be public. This gives an opportunity for people to affect the mocking behaviour adversely. I would recommend a review of the public classes and make any ones that are not required, internal.
- Namespaces. The one namespace I am having a bit of issue with is the Syntax namespace. Not sure what it means and also it makes sense for the Mock class to reside in HyperMock.Universal not HyperMock.Universal.Syntax. The others make more sense though.
- Post-fork changes. Since you forked I have added support for events and also an overload for the Throws behaviour to allow callers to specify an instance of an exception. These both would need to be added. The events might take a little more effort given the internal changes.
- Non-generic Mock creation. In my original version of the Mock class, I added support for creating the Mock from a type not just a generic class. It tends to be a good idea to support both as it makes it easier to automate creation of mock instances by reflection (if desired). This might be a bit of an edge case mind!
- Non-universal option. I always had in mind to look at creating a version of HyperMock that worked outside of just UWP projects. Given the great changes you have made, I can see this being an easy win. If we abstract the proxy generation we can hopefully have a base library that will work across BCL and UWP platforms. 

Let me know what you think and we can look at progressing from here.

Kind regards
Steve

**Comment by nelsonghezzi** [comment]
Hi Steve,

Good news you are subdividing each project in their own repo! I was going to suggest that too, for the same reasons pointed by @cristi-badila.

I'm working on another syntax improvement in my fork. Was going to submit a PR to discuss it (I'm not totally sure if the names I've picked are the best) and ask what you think about it, but maybe it would be better to wait until you create the repo for HyperMock.

I also thought about some other refactors / internal reorganizations of the codebase as I was learning more about it, so I'll keep track of the ideas shared here.

This library fills an empty spot, being the only one for this kind of projects, so I'm glad to see more people involved in making it evolve and improve, and the things discussed here.

Regards.

**Comment by steve-hyperbolic** [comment]
Hi Nelson

I am currently sorting out some of the internal parts to make it a bit cleaner. Should be done with it in a day or two. Feel free to continue with your syntax changes etc in your fork. I can review it and port the code appropriately into the new HyperMock repo.

Once it's up and running it will make things like forking/PRs a bit easier.

I have already moved the HyperIoC out but just need to finish the internal bits for HyperMock before doing likewise.

Cheers
Steve

**Comment by steve-hyperbolic** [comment]
Hello

Just to let you know I have now moved HyperMock into its own repo:

[HyperMock](https://github.com/steve-hyperbolic/hypermock)

We have an opportunity to extend this framework to cover everything needed for mocking tests.

I am thinking that what we can do is create a 2.0 version to build upon that can contain any breaking changes such as Cristian's suggestions to improve the API (as there is some great stuff there) along with other missing parts such as event and extra overloads for fuzzy matching (for example regex). I am also keen to do what I have done with HyperIoC and make it available for all .NET platforms. See [HyperIoC](https://github.com/steve-hyperbolic/hyperioc) for more details.

I also want to address the documentation too, in much the same way as HyperIoC to make it more comprehensive and clearer.

In the interim we might need to maintain the current version until a point at which the 2.0 version is ready.

My plan would be to create a release 1 and release 2 branches from master with release 2 being the 2.0 release work.

Let me know your thoughts and I can setup things accordingly.

Regards
Steve

**Comment by nelsonghezzi** [comment]
Agree with having separate branches for v1 and a future v2, leaving master for what it's "production ready" (i.e. the commit from which the Nuget package is generated). All contrib PRs should be made against v1 or v2 branches.

I'm still thinking about appropriate names for these. Maybe v1 = develop/v1, v2 = develop/v2 ? or just simply v1 and v2 (or release1 and release2) ?

Perhaps I would have taken a different approach for the repo migration. At first glance, it seems like it wouldn't been path conflicts, so whenever possible, making a series of patches for the HyperMock commits would have preserved some of the history.

+1 for docs too. Have to take a look to the HyperIoC one.

Cheers!
